### PR TITLE
Implement + unit test explore_clustering

### LINKED ISSUE #27: explore_clustering - write test cases, code iteratively, handle exceptions
## Function: `explore_clustering`
- [ ] **Write test cases and code iteratively:**

- Before writing any code, revisit your function specifications from last week and revise them if needed. As you do this, think critically about what function inputs you expect from the user and what your function is supposed to do (e.g., return).

- Write a unit test function for each specified function, which is named after the function being tested (e.g., if the function is named foo then the unit test function is named test_foo). The unit test function should test 3 to 5 edge cases to ensure that the function returns what is expected to the user. When writing your tests, build them around your function specifications and requirements.

- After writing your unit tests, write the internal code for your Python functions.

- Write function code and tests in close temporal proximity with each other. The process of writing function code and test cases will be iterative; there will be several rounds of writing tests → function improvement → writing more tests → function improvement → writing more tests → function improvement, etc.

- Make sure that the tests cover all branches (e.g., if your function has an if statement, you should have tests where the if statement is true, and where the if statement is false.

- As you develop your code and test cases, update your code documentation so that it makes sense with any of the changes you have made. Do not forget to also document your test functions.

- Check whether the code passes the unit tests you have written using the pytest package.

- [ ] **Exception handling**:

- The function should be written defensively. That is, it should handle incorrect input and errors detected during execution via throwing exceptions with useful error messages, and there should be tests to confirm that the exceptions result in the expected behaviour of the function.

- [ ] **Planning and organizing your work**:

- Any ideas/notes that you generate that are related to your assigned function should be recorded in the corresponding issue you created in milestone 1. The issue should be closed by the end of milestone 2.

- If you think of new ideas or test cases and you do not have time to implement them in this milestone, create an issue for them which you can complete during the next milestone.

### PR DESCRIPTION
**Pull request by lephanthuymai**
Hi Aditya et. al,
Please help to review my PR with changes below:
- Implement and write unit tests for explore_clustering
- Implement and write unit tests for explore_KMeans_clustering
- Implement and write unit tests for explore_DBSCAN_clustering
- Create eda_playround.ipynb to give an example of how we use the functions for EDA tasks on a dataset

I have run all the tests locally, they all passed.

Notes:
- I'm using [menus.csv ](https://www.kaggle.com/mcdonalds/nutrition-facts) as the test dataset, there are numeric, categorical and text columns, should be fine to utilize all of our EDA functions
- There are issues with testing matplotlib plots, I have to use `matplotlib.use("Agg")` to disable opening plot windows in the test file, it causes some texts not generated for the plot objects, hence I can only do some simple tests for these plots. There are some other ideas for testing this kind of plot: https://stackoverflow.com/questions/27948126/how-can-i-write-unit-tests-against-code-that-uses-matplotlib, which I find quite complex. If you have better ideas on how to test them, please let me know.

Thanks.

### CODE REVIEW

File: datascience_eda/datascience_eda.py [hunk 1]
```diff
@@ -7,28 +7,82 @@
...
+    -------
+    Exception
+        throw an exception if the metric is invalid
+    """
+    if not m in ["cityblock", "cosine", "euclidean", "l1", "l2", "manhattan"]:
+        raise Exception(f"Invalid distance metric: {m}")
+
+
+def _verify_numeric_cols(df, num_cols):
+    """check if numeric columns are valie
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]: Could you please re check the spelling of valid in documentation 
```suggestion
    """check if numeric columns are valid
```

File: datascience_eda/datascience_eda.py [hunk 2]
```diff
@@ -41,31 +95,71 @@ def get_clustering_default_hyperparameters():
...
-    df, hyperparameter_dict=get_clustering_default_hyperparameters()
+    df, numeric_cols=None, hyperparameter_dict=get_clustering_default_hyperparameters()
 ):
     """fit and plot K-Means, DBScan clustering algorithm on the dataset
 
     Parameters
     ----------
     df : pandas.DataFrame
-        the dataset (X)
+        the dataset (X), should already transformed with StandardScaler
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]:
```suggestion
        the dataset (X), should already be transformed with StandardScaler()
```

File: datascience_eda/datascience_eda.py [hunk 3]
```diff
@@ -41,31 +95,71 @@ def get_clustering_default_hyperparameters():
...
+    numeric_cols: list, optional
+        a list of numeric columns used for clustering, by default None, will be assigned with all numeric columns
     hyperparameter_dict : dict, optional
-        the distance metric and hyperparameters to be used in the clustering algorithm, by default get_clustering_default_hyperparameters()
+        the hyperparameters to be used in the clustering algorithms, by default get_clustering_default_hyperparameters()
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]: How would it be if the values of hyper parameters are directly mentioned in place of/after function name so that the user need not go and check get_clustering_default_hyperamaeters() function?

File: datascience_eda/datascience_eda.py [hunk 4]
```diff
@@ -74,31 +168,293 @@ def explore_clustering(
 
     Examples
     -------
-    >>> explore_clustering(X)
+    >>> original_df = pd.read_csv(currentdir + "/data/menu.csv")
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]: the variable currentdir could be defined before original_df

File: datascience_eda/datascience_eda.py [hunk 5]
```diff
@@ -74,31 +168,293 @@ def explore_clustering(
...
+    if not ("DBSCAN" in hyperparameter_dict):
+        raise Exception("Expecting DBSCAN hyperparams.")
+
+    kmeans_params = hyperparameter_dict["KMeans"]
+    if not ("n_clusters" in kmeans_params):
+        raise Exception("Expecting n_clusters in KMeans' hyperparams.")
+
+    dbscan_params = hyperparameter_dict["DBSCAN"]
+    if not ("eps" in dbscan_params):
+        raise Exception("Excepting eps in DBSCAN's hyperparams.")
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]:
```suggestion
        raise Exception("Expecting eps in DBSCAN's hyperparams.")
```

File: datascience_eda/datascience_eda.py [hunk 6]
```diff
@@ -74,31 +168,293 @@ def explore_clustering(
...
+    metric = dbscan_params["distance_metric"]
+    _verify_distance_metric(metric)
+
+    # endregion
+
+    print("***********************")
+    print("EXPLORE CLUSTERING")
+    print("***********************")
+
+    kmeans_plots = explore_KMeans_clustering(
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]: explore_KMeans_clustering is defined after explore_clustering. How would it be if explore_KMeans_clustering is defined before explore_clustering so that there would be a flow while referring to the code?

File: datascience_eda/datascience_eda.py [hunk 7]
```diff
@@ -74,31 +168,293 @@ def explore_clustering(
...
+        df,
+        num_cols=numeric_cols,
+        n_clusters=kmeans_params["n_clusters"],
+    )
+
+    dbscan_plots = explore_DBSCAN_clustering(
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]: explore_DBSCAN_clustering is defined after explore_clustering. How would it be if explore_DBSCAN_clustering is defined before explore_clustering so that there would be a flow while referring to the code?

File: datascience_eda/datascience_eda.py [hunk 8]
```diff
@@ -74,31 +168,293 @@ def explore_clustering(
...
-    # ---clustering with DBSCAN---------
-    # get eps and min_samples from hyperparameter_dict["DBSCAN"]
+    Returns
+    -------
+    dict
+        a dictionary with key=type of plot, value=list of plots
+
+    Examples
+    -------
+    >>> original_df = pd.read_csv(currentdir + "/data/menu.csv")
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]: the variable currentdir could be defined before original_df

File: datascience_eda/datascience_eda.py [hunk 9]
```diff
@@ -74,31 +168,293 @@ def explore_clustering(
...
+    -------
+    Tuple
+        list
+            a list of n_clusters values returned by DBSCAN models
+        dict
+            a dictionary with key=type of plot, value=list of plots
+
+    Examples
+    -------
+    >>> original_df = pd.read_csv(currentdir + "/data/menu.csv")
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]: the variable currentdir could be defined before original_df

File: datascience_eda/datascience_eda.py [hunk 10]
```diff
@@ -74,31 +168,293 @@ def explore_clustering(
...
+            s_visualizer.fit(x)  # Fit the data to the visualizer
+            s_visualizer.show()
+            plt.clf()
+            plt.close()
+            silhouette_plots.append(s_visualizer)
+        else:
+            silhouette_plots.append(None)
+
+        # PCA plots
+        if include_silhouette:
```
Comment (adibns) [inline datascience_eda/datascience_eda.py]: Could you please check if this if include_silhouette is if include_pca?
```suggestion
        if include_pca:
```

File: datascience_eda/datascience_eda.py [hunk 11]
```diff
@@ -74,31 +168,293 @@ def explore_clustering(
...
+    s_plots = []
+    pca_plots = []
+
+    print("------------------------")
+    print("DBSCAN CLUSTERING")
+    print("------------------------")
+
+    for e in eps:
+        for ms in min_samples:
+            dbscan = DBSCAN(eps=e, min_samples=ms, metric=metric)
```
Comment (adibns) [inline datascience_eda/datascience_eda.py:333]: Could you please let me know if the there would be an error message thrown when metric is not of string type?

**Review by adibns — COMMENTED** [review]
Great work!

File: datascience_eda/datascience_eda.py (hunk 11 above)
Comment (lephanthuymai) [reply datascience_eda/datascience_eda.py:333]: the metric is verified earlier in the function

**Comment by lephanthuymai** [comment]
Hi @adibns ,
After some discussions with Tiff, Peter (TA), and Varada, I have updated the unit tests for verifying plots.
The tests passed and the EDA notebook renders fine.
Sorry that it will take you some time to review my commits again.
Thank you for your help.
Best,
Mai
